src/plugins/stake-plugins/generic/policyAdapters/TarotPoolAdaptor.ts

```diff
@@ -0,0 +1,491 @@
+            })
+        )
+      }
+      const token1BorrowableContract = TarotBorrowable__factory.connect(token1BorrowableContractAddress, provider)
+      const token1BorrowableAllowance = await token1BorrowableContract.allowance(walletAddress, tarotRouterContractAddress)
+      if (token1BorrowableAllowance.lt(ethers.constants.MaxUint256)) {
+        txs.push(
+          async () =>
+            await token1BorrowableContract.populateTransaction.borrowApprove(tarotRouterContractAddress, ethers.constants.MaxUint256, {
+              gasPrice,
+              gasLimit: '1000000',
+              customData: {
+                name: metadataName,
+                category: 'Expense:Fees',
+                notes: `Approve bTarot ${metadataPoolAssetName} liquidity pool contract`
+              }
+            })
+        )
+      }
+
+      // do the leverage
+      txs.push(async (prevTx?: ethers.providers.TransactionResponse) => {
+        if (prevTx != null) {
+          await prevTx.wait(1)
+        }
+        return await tarotRouterContract.populateTransaction.leverage(
+          poolContractAddress,
+          token0Amount,
+          token1Amount,
+          amountToken0Min,
+          amountToken1Min,
+          walletAddress,
+          deadline,
+          '0x',
+          '0x',
+          {
+            gasPrice,
+            gasLimit: '10000000',
+            customData: {
+              name: metadataName,
+              category: 'Transfer:Staking',
+              notes: `Leverage ${metadataPoolAssetName} liquidity pool contract`
+            }
+          }
+        )
+      })
+
+      // Calculate the stake asset native amounts:
+      const allocations: QuoteAllocation[] = [
+        {
+          allocationType: 'stake',
+          pluginId: policyConfig.stakeAssets[0].pluginId,
+          currencyCode: policyConfig.stakeAssets[0].currencyCode,
+          nativeAmount: token0Amount.toString()
+        },
+        {
+          allocationType: 'stake',
+          pluginId: policyConfig.stakeAssets[1].pluginId,
+          currencyCode: policyConfig.stakeAssets[1].currencyCode,
+          nativeAmount: token1Amount.toString()
+        }
+      ]
+
+      return await prepareChangeQuote(walletSigner, txs, allocations)
+    },
+    fetchUnstakeQuote: async (wallet: EdgeCurrencyWallet, requestAssetId: AssetId, nativeAmount: string): Promise<ChangeQuote> => {
+      const { gasPrice, txs, walletAddress, walletSigner } = await workflowUtils(wallet)
+      const collateralContract = TarotCollateral__factory.connect(collateralContractAddress, provider)
+      const tarotRouterContract = TarotRouter__factory.connect(tarotRouterContractAddress, provider)
+
+      const tarot = tarotUtils(adapterConfig, provider, walletAddress)
+
+      // approve collateral token
+      const collateralAllowance = await collateralContract.allowance(walletAddress, tarotRouterContractAddress)
+      if (collateralAllowance.lt(ethers.constants.MaxUint256)) {
```

Comment:
We only need to approve the amount once and tarot.to uses the max

Reply:
Not sure I understand. If you give a max allowance that means you're allowing for more than your entire balance. If the SC moves your entire balance, they still have an allowance of `MaxUint256 - balance`. When you get more collateral, you're balance most likely will be less then this remaining allowance and so we don't need to make another approve call. But, because you're comparing the allowance with `MaxUint235`, which is greater than `MaxUint256 - n` where n > 0, then you're always doing an approve call. 

Perhaps you want `lt(balance)` instead?

Reply:
Resolve with the total supply trick